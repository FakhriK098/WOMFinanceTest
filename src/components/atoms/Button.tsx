import React from 'react';
import {
  ActivityIndicator,
  Pressable,
  StyleSheet,
  View,
  PressableStateCallbackType,
} from 'react-native';
import Text from './Text';
import { colors } from '@theme/colors';
import { spacing } from '@theme/spacing';
import { radius } from '@theme/radius';
import { ButtonAtomProps } from '@typings/components';

const Button = ({
  title,
  loading,
  disabled,
  style,
  ...rest
}: ButtonAtomProps) => {
  const isDisabled = disabled || loading;
  const computedStyle = (state: PressableStateCallbackType) => {
    const propStyle = typeof style === 'function' ? style(state) : style;
    return [styles.base, isDisabled && styles.disabled, propStyle];
  };
  return (
    <Pressable
      accessibilityRole="button"
      style={computedStyle}
      disabled={isDisabled}
      {...rest}
    >
      <View style={styles.row}>
        {loading && (
          <ActivityIndicator
            color={colors.white}
            style={{ marginRight: spacing.sm }}
          />
        )}
        <Text color={colors.white} weight="bold">
          {title}
        </Text>
      </View>
    </Pressable>
  );
};

const styles = StyleSheet.create({
  base: {
    backgroundColor: colors.primary,
    paddingVertical: spacing.md,
    borderRadius: radius.md,
    alignItems: 'center',
  },
  disabled: {
    opacity: 0.6,
  },
  row: {
    flexDirection: 'row',
    alignItems: 'center',
  },
});

export default React.memo(Button);
