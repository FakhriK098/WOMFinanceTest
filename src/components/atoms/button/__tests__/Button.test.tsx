import React from 'react';
import renderer, { act } from 'react-test-renderer';
import { ActivityIndicator } from 'react-native';

import Button from '../Button';

describe('Button atom', () => {
  it('matches snapshot (default)', () => {
    let tr: renderer.ReactTestRenderer;
    act(() => {
      tr = renderer.create(<Button title="Snapshot" />);
    });
    expect(tr!.toJSON()).toMatchSnapshot();
    act(() => tr!.unmount());
  });

  it('renders the title text', () => {
    let testRenderer: renderer.ReactTestRenderer;
    act(() => {
      testRenderer = renderer.create(<Button title="Submit" />);
    });
    const json = testRenderer!.toJSON();
    const hasSubmit = (node: any): boolean => {
      if (!node) return false;
      if (typeof node === 'string') return node === 'Submit';
      const children = node.children || [];
      if (Array.isArray(children)) return children.some(hasSubmit);
      return false;
    };
    expect(hasSubmit(json)).toBe(true);
    act(() => {
      testRenderer!.unmount();
    });
  });

  it('disables when disabled prop is true', () => {
    let testRenderer: renderer.ReactTestRenderer;
    act(() => {
      testRenderer = renderer.create(<Button title="Tap" disabled />);
    });
    const pressable = testRenderer!.root.find(
      inst => inst.props && inst.props.accessibilityRole === 'button',
    );
    expect(pressable.props.disabled).toBe(true);
    act(() => {
      testRenderer!.unmount();
    });
  });

  it('shows loader and disables when loading', () => {
    let testRenderer: renderer.ReactTestRenderer;
    act(() => {
      testRenderer = renderer.create(<Button title="Loading" loading />);
    });
    // Snapshot when loading
    expect(testRenderer!.toJSON()).toMatchSnapshot();
    // ActivityIndicator should be present
    const loaders = testRenderer!.root.findAllByType(ActivityIndicator);
    expect(loaders.length).toBe(1);
    // Pressable should be disabled while loading
    const pressable = testRenderer!.root.find(
      inst => inst.props && inst.props.accessibilityRole === 'button',
    );
    expect(pressable.props.disabled).toBe(true);
    act(() => {
      testRenderer!.unmount();
    });
  });
});
