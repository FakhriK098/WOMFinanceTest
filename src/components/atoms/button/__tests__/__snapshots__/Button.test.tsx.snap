// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button atom matches snapshot (default) 1`] = `
<View
  accessibilityRole="button"
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    [
      {
        "alignItems": "center",
        "backgroundColor": "#4F46E5",
        "borderRadius": 12,
        "paddingVertical": 14,
      },
      undefined,
      undefined,
    ]
  }
>
  <View
    style={
      {
        "alignItems": "center",
        "flexDirection": "row",
      }
    }
  >
    <Text
      style={
        [
          {
            "color": "#FFFFFF",
            "fontSize": 16,
            "fontWeight": "700",
          },
          undefined,
        ]
      }
    >
      Snapshot
    </Text>
  </View>
</View>
`;

exports[`Button atom shows loader and disables when loading 1`] = `
<View
  accessibilityRole="button"
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": true,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    [
      {
        "alignItems": "center",
        "backgroundColor": "#4F46E5",
        "borderRadius": 12,
        "paddingVertical": 14,
      },
      {
        "opacity": 0.6,
      },
      undefined,
    ]
  }
>
  <View
    style={
      {
        "alignItems": "center",
        "flexDirection": "row",
      }
    }
  >
    <ActivityIndicator
      color="#FFFFFF"
      style={
        {
          "marginRight": 8,
        }
      }
    />
    <Text
      style={
        [
          {
            "color": "#FFFFFF",
            "fontSize": 16,
            "fontWeight": "700",
          },
          undefined,
        ]
      }
    >
      Loading
    </Text>
  </View>
</View>
`;
